core/stat/base/slidingwindow_metric.go

```diff
@@ -0,0 +1,118 @@
+package base
+
+import (
+	"fmt"
+	"github.com/sentinel-group/sentinel-golang/core/base"
+	"github.com/sentinel-group/sentinel-golang/util"
+	"math"
+)
+
+// The basic metric struct,
+// SlidingWindowMetric doesn't store any data and is the wrapper of BucketLeapArray to adapt to different internal window
+// SlidingWindowMetric is used for SentinelRules and BucketLeapArray is used for monitor
+// BucketLeapArray is per resource, and SlidingWindowMetric support only read operation.
+type SlidingWindowMetric struct {
+	windowLengthInMs uint32
+	sampleCount      uint32
+	intervalInMs     uint32
+	real             *BucketLeapArray
+}
+
+// It must pass the parameter point to the real storage entity
+func NewSlidingWindowMetric(sampleCount, intervalInMs uint32, real *BucketLeapArray) *SlidingWindowMetric {
+	if real == nil || intervalInMs <= 0 || sampleCount <= 0 {
+		panic(fmt.Sprintf("Illegal parameters,intervalInMs=%d,sampleCount=%d,real=%+v.", intervalInMs, sampleCount, real))
+	}
+
+	if intervalInMs%sampleCount != 0 {
+		panic(fmt.Sprintf("Invalid parameters, intervalInMs is %d, sampleCount is %d.", intervalInMs, sampleCount))
+	}
+	winLengthInMs := intervalInMs / sampleCount
+
+	parentIntervalInMs := real.IntervalInMs()
+	parentWindowLengthInMs := real.WindowLengthInMs()
+
+	// winLengthInMs of BucketLeapArray must be divisible by winLengthInMs of SlidingWindowMetric
+	// for example: winLengthInMs of BucketLeapArray is 500ms, and winLengthInMs of SlidingWindowMetric is 2000ms
+	// for example: winLengthInMs of BucketLeapArray is 500ms, and winLengthInMs of SlidingWindowMetric is 500ms
+	if winLengthInMs%parentWindowLengthInMs!=0 {
+		panic(fmt.Sprintf("BucketLeapArray's WindowLengthInMs(%d) is not divisible by SlidingWindowMetric's WindowLengthInMs(%d).", parentWindowLengthInMs, winLengthInMs))
+	}
+
+	if intervalInMs > parentIntervalInMs {
+		// todo if SlidingWindowMetric's intervalInMs is greater than BucketLeapArray.
+		panic(fmt.Sprintf("The interval(%d) of SlidingWindowMetric is greater than parent BucketLeapArray(%d).", intervalInMs, parentIntervalInMs))
+	}
+
+	// 10 * 1000 ms == parent
+	if parentIntervalInMs%intervalInMs != 0 {
+		panic(fmt.Sprintf("SlidingWindowMetric's intervalInMs(%d) is not divisible by real BucketLeapArray's intervalInMs(%d).", intervalInMs, parentIntervalInMs))
+	}
+
+	return &SlidingWindowMetric{
+		windowLengthInMs: winLengthInMs,
+		sampleCount:      sampleCount,
+		intervalInMs:     intervalInMs,
+		real:             real,
+	}
+}
+
+// Get the [start time, end time) of SlidingWindowMetric for now.
+func (m *SlidingWindowMetric) getTimeInterval(now uint64) (start, end uint64) {
+	//
+	timeId := now / uint64(m.windowLengthInMs)
+	idx :=  uint32(timeId) % m.sampleCount
+	bucketStartTime := calculateStartTime(now, m.windowLengthInMs)
+
+	start = bucketStartTime - uint64(idx * m.windowLengthInMs)
+	end = start + uint64(m.intervalInMs)
```

Comment:
I'm confused by the approach to calculating the condition range. As the latest idx is consistent with current time, maybe the `end` should be yielded first? Here is my code snippet (may not cover all scenarios):

```go
bucketStartTime := calculateStartTime(now, m.real.WindowLengthInMs())

// bucket range (startTime): [start, end]
end = bucketStartTime
start = end - uint64(m.intervalInMs)
```

My test case: inner LeapArray(ia=2s, sa=20), the wrapper Metric(im=2s, sm=10), current timestamp is 1578476668307, then the range of the **inner bucket start time** should be [1578476666300,1578476668300]